packages/vm/src/compatability.rs

```diff
@@ -46,66 +46,58 @@ pub fn check_wasm(wasm_code: &[u8]) -> Result<()> {
-                missing, REQUIRED_EXPORTS
-            ));
-    }
-    Ok(())
-}
-
-fn check_wasm_imports(module: &Module) -> Result<()> {
-    if let Some(missing) = find_missing_import(module, SUPPORTED_IMPORTS) {
-        return make_validation_err(format!(
-                "Wasm contract requires unsupported import: \"{}\". Imports supported by VM: {:?}. Contract version too new for this VM?",
-                missing, SUPPORTED_IMPORTS
-            ));
-    }
-    Ok(())
-}
-
-fn find_missing_export(module: &Module, required_exports: &[&str]) -> Option<String> {
-    let available_exports: Vec<String> = match module.export_section() {
-        Some(export_section) => Vec::from(export_section.entries())
+    let available_exports: Vec<String> = module.export_section().map_or(vec![], |export_section| {
+        export_section
+            .entries()
             .iter()
             .map(|entry| entry.field().to_string())
-            .collect(),
-        None => vec![],
-    };
+            .collect()
+    });
 
-    for required_export in required_exports {
+    for required_export in REQUIRED_EXPORTS {
         if !available_exports.iter().any(|x| x == required_export) {
-            return Some(String::from(*required_export));
+            return make_validation_err(format!(
+                "Wasm contract doesn't have required export: \"{}\". Exports required by VM: {:?}. Contract version too old for this VM?",
+                required_export, REQUIRED_EXPORTS
+            ));
         }
     }
-    None
+    Ok(())
 }
 
 /// Checks if the import requirements of the contract are satisfied.
 /// When this is not the case, we either have an incompatibility between contract and VM
 /// or a error in the contract.
-fn find_missing_import(module: &Module, supported_imports: &[&str]) -> Option<String> {
-    let required_imports: Vec<String> = match module.import_section() {
-        Some(import_section) => Vec::from(import_section.entries())
-            .iter()
-            .map(|entry| format!("{}.{}", entry.module(), entry.field()))
-            .collect(),
-        None => vec![],
-    };
+fn check_wasm_imports(module: &Module) -> Result<()> {
+    let required_imports: Vec<ImportEntry> = module
+        .import_section()
+        .map_or(vec![], |import_section| import_section.entries().to_vec());
 
     for required_import in required_imports {
-        if !supported_imports.contains(&required_import.as_str()) {
-            return Some(required_import);
+        let full_name = format!("{}.{}", required_import.module(), required_import.field());
+        if !SUPPORTED_IMPORTS.contains(&full_name.as_str()) {
+            return make_validation_err(format!(
+                "Wasm contract requires unsupported import: \"{}\". Imports supported by VM: {:?}. Contract version too new for this VM?",
+                full_name, SUPPORTED_IMPORTS
+            ));
         }
+
+        match required_import.external() {
```

Comment:
This is the key check beyond cleanup, right? Looks good. :+1:

Just a thought, is it possible to view function signature there? Probably overkill (and for another PR), but curious if we could enforce number of args and return value here.

Reply:
> is it possible to view function signature there?

I think so. Near is doing that here: https://github.com/nearprotocol/nearcore/blob/4bc4b74aeaab3231866ddb7de822e8892a13d9dc/runtime/near-vm-runner/src/prepare.rs#L76-L77 However, I did not look into it in detail.

Reply:
> This is the key check beyond cleanup, right?

Exactly. The rest is refactoring of the methods. I found it easier to only return `Result<()>` and stop at the first failed check.